### ISSUE TITLE
Jetty 12 Core SecurityHandler relies on ContextHandler

### ISSUE THREAD

**Issue by lachlan-roberts** [eclipse/jetty.project#11881]
**Jetty version(s)**
12.0.x

**Description**
The core `SecurityHandler` currently depends on being behind a `ContextHandler`. It implements the method:

```java
public static SecurityHandler getCurrentSecurityHandler()
{
    ContextHandler contextHandler = ContextHandler.getCurrentContextHandler();
    if (contextHandler != null)
        return contextHandler.getDescendant(SecurityHandler.class);
    return null;
}
```

Which only works if it has a `ContextHandler`.

I think we should either; implement a way to get the current `SecurityHandler` without relying on `ContextHandler` OR we enforce that we are behind a `ContextHandler` during the `SecurityHandler.doStart()`.

**Comment by gregw** [comment]
It should be easy enough to add some code that if contextHandler is null, then to ask the server for its descendants security handlers.

Who calls this method and why?

**Comment by lachlan-roberts** [comment]
It is used by `UserAuthenticationSucceeded#logout` and `SessionAuthentication#readObject`.

> It should be easy enough to add some code that if contextHandler is null, then to ask the server for its descendants security handlers.

Is there a static method to get the server somewhere? I couldn't find one.

**Comment by lachlan-roberts** [comment]
> It should be easy enough to add some code that if contextHandler is null, then to ask the server for its descendants security handlers.

I don't think this is sufficient, as it would be possible to setup a handler structure where it is not possible to determine which `SecurityHandler` to use.

Here are the 3 main usages of `getCurrentSecurityHandler()` and some possible solutions:

- `DefaultAuthenticatorFactory` - an additional method could be added on the `Authenticator.Configuration` interface to give a reference to the `SecurityHandler`.
- `LoginAuthenticator.logout()` - this has references to the request/response objects, so the `SecurityHandler` could store itself on a request attribute which could then be obtained during `logout()`.
- `SessionAuthentication.readObject()` - when the `SessionAuthentication` is deserialized uses the static method to get the `SecurityHandler` and re-validates itself with the `LoginService`, so we would need to remove this logic from `readObject()` and have an additional step initiated from the `Authenticator`s where if they retrieve a `SessionAuthentication` they must also verify it with the `LoginService`.

My main concern is the change for the last point requires the `Authenticator`s to be modified to use this new pattern, and could cause issues for any user who has implemented their own `Authenticator` which uses `SessionAuthentication`.

So maybe it would be preferable to just enforce that `SecurityHandler` exists behind a `ContextHandler` during `SecurityHandler.doStart()`.
